velox/functions/sparksql/Hash.cpp

```diff
@@ -101,7 +104,7 @@ class SparkVectorHasher {
   virtual ~SparkVectorHasher() = default;
 
   // Compute the hash value of input vector at index.
-  ReturnType hashAt(vector_size_t index, SeedType seed) {
+  virtual ReturnType hashAt(vector_size_t index, SeedType seed) {
```

Comment:
Do we have a benchmark to measure this? The hot function change to virtual function. I would suggest to not change here, because UnknownType is not a common used type, we don't need to do more optimization for it.

Reply:
@boneanxs We have a benchmark here https://github.com/facebookincubator/velox/blob/main/velox/functions/sparksql/benchmarks/HashBenchmark.cpp

Reply:
Looks there's no much difference if `hashAt` is virtual, but I agree with you that UnknownType doesn't need extra optimization that could affect others performance, So I remove the `virtual` here. Pls review again.

`hashAt` is Virtual:

```c++
============================================================================
[...]hmarks/ExpressionBenchmarkBuilder.cpp     relative  time/iter   iters/s
============================================================================
hash#ARRAY<MAP<INTEGER,VARCHAR>>#0%nulls##hash             23.71ms     42.17
hash#ARRAY<MAP<INTEGER,VARCHAR>>#0%nulls##xxhas            23.00ms     43.48
----------------------------------------------------------------------------
hash#ARRAY<MAP<INTEGER,VARCHAR>>#25%nulls##hash            17.38ms     57.53
hash#ARRAY<MAP<INTEGER,VARCHAR>>#25%nulls##xxha            17.55ms     56.98
----------------------------------------------------------------------------
hash#BIGINT#0%nulls##hash                                   2.80ms    357.56
hash#BIGINT#0%nulls##xxhash64                               2.49ms    402.23
----------------------------------------------------------------------------
hash#BIGINT#25%nulls##hash                                  2.88ms    347.51
hash#BIGINT#25%nulls##xxhash64                              2.86ms    349.89
----------------------------------------------------------------------------
hash#DECIMAL(18, 6)#0%nulls##hash                           2.80ms    357.41
hash#DECIMAL(18, 6)#0%nulls##xxhash64                       2.47ms    404.26
----------------------------------------------------------------------------
hash#DECIMAL(18, 6)#25%nulls##hash                          2.88ms    347.55
hash#DECIMAL(18, 6)#25%nulls##xxhash64                      2.86ms    349.49
----------------------------------------------------------------------------
hash#DECIMAL(38, 6)#0%nulls##hash                           7.92ms    126.29
hash#DECIMAL(38, 6)#0%nulls##xxhash64                       8.76ms    114.12
----------------------------------------------------------------------------
hash#DECIMAL(38, 6)#25%nulls##hash                          6.98ms    143.29
hash#DECIMAL(38, 6)#25%nulls##xxhash64                      7.44ms    134.48
----------------------------------------------------------------------------
hash#DOUBLE#0%nulls##hash                                   3.12ms    320.21
hash#DOUBLE#0%nulls##xxhash64                               2.50ms    399.46
----------------------------------------------------------------------------
hash#DOUBLE#25%nulls##hash                                  3.14ms    318.82
hash#DOUBLE#25%nulls##xxhash64                              2.88ms    347.35
----------------------------------------------------------------------------
hash#REAL#0%nulls##hash                                     1.71ms    583.52
hash#REAL#0%nulls##xxhash64                                 2.55ms    391.95
----------------------------------------------------------------------------
hash#REAL#25%nulls##hash                                    2.76ms    362.86
hash#REAL#25%nulls##xxhash64                                2.71ms    369.33
----------------------------------------------------------------------------
hash#ROW<f_map:MAP<INTEGER,VARCHAR>,f_array:ARR            38.57ms     25.93
hash#ROW<f_map:MAP<INTEGER,VARCHAR>,f_array:ARR            37.05ms     26.99
----------------------------------------------------------------------------
hash#ROW<f_map:MAP<INTEGER,VARCHAR>,f_array:ARR            27.27ms     36.67
hash#ROW<f_map:MAP<INTEGER,VARCHAR>,f_array:ARR            27.04ms     36.99
----------------------------------------------------------------------------
hash#SMALLINT#0%nulls##hash                                 1.66ms    603.96
hash#SMALLINT#0%nulls##xxhash64                             2.36ms    423.40
----------------------------------------------------------------------------
hash#SMALLINT#25%nulls##hash                                2.51ms    399.10
hash#SMALLINT#25%nulls##xxhash64                            2.47ms    404.36
----------------------------------------------------------------------------
hash#TIMESTAMP#0%nulls##hash                                4.71ms    212.14
hash#TIMESTAMP#0%nulls##xxhash64                            4.67ms    214.20
----------------------------------------------------------------------------
hash#TIMESTAMP#25%nulls##hash                               4.12ms    242.89
hash#TIMESTAMP#25%nulls##xxhash64                           4.06ms    246.35
----------------------------------------------------------------------------
hash#TINYINT#0%nulls##hash                                  1.61ms    621.65
hash#TINYINT#0%nulls##xxhash64                              2.26ms    442.19
----------------------------------------------------------------------------
hash#TINYINT#25%nulls##hash                                 2.49ms    401.84
hash#TINYINT#25%nulls##xxhash64                             2.39ms    418.30
----------------------------------------------------------------------------
hash#UNKNOWN#0%nulls##hash                                  1.33ms    754.70
hash#UNKNOWN#0%nulls##xxhash64                              1.32ms    755.79
----------------------------------------------------------------------------
hash#UNKNOWN#25%nulls##hash                                 1.33ms    753.16
hash#UNKNOWN#25%nulls##xxhash64                             1.30ms    772.15
----------------------------------------------------------------------------
hash#VARCHAR#0%nulls##hash                                 10.98ms     91.11
hash#VARCHAR#0%nulls##xxhash64                              9.91ms    100.86
----------------------------------------------------------------------------
hash#VARCHAR#25%nulls##hash                                 9.21ms    108.56
hash#VARCHAR#25%nulls##xxhash64                             8.40ms    119.01
----------------------------------------------------------------------------
```

`hashAt` is not Virtual:

```c++
============================================================================
[...]hmarks/ExpressionBenchmarkBuilder.cpp     relative  time/iter   iters/s
============================================================================
hash#ARRAY<MAP<INTEGER,VARCHAR>>#0%nulls##hash             24.02ms     41.63
hash#ARRAY<MAP<INTEGER,VARCHAR>>#0%nulls##xxhas            23.22ms     43.06
----------------------------------------------------------------------------
hash#ARRAY<MAP<INTEGER,VARCHAR>>#25%nulls##hash            17.28ms     57.86
hash#ARRAY<MAP<INTEGER,VARCHAR>>#25%nulls##xxha            16.60ms     60.25
----------------------------------------------------------------------------
hash#BIGINT#0%nulls##hash                                   2.84ms    351.85
hash#BIGINT#0%nulls##xxhash64                               2.48ms    403.17
----------------------------------------------------------------------------
hash#BIGINT#25%nulls##hash                                  2.92ms    343.02
hash#BIGINT#25%nulls##xxhash64                              2.62ms    382.22
----------------------------------------------------------------------------
hash#DECIMAL(18, 6)#0%nulls##hash                           2.83ms    353.96
hash#DECIMAL(18, 6)#0%nulls##xxhash64                       2.47ms    404.40
----------------------------------------------------------------------------
hash#DECIMAL(18, 6)#25%nulls##hash                          2.90ms    344.70
hash#DECIMAL(18, 6)#25%nulls##xxhash64                      2.60ms    383.90
----------------------------------------------------------------------------
hash#DECIMAL(38, 6)#0%nulls##hash                           7.93ms    126.14
hash#DECIMAL(38, 6)#0%nulls##xxhash64                       8.51ms    117.49
----------------------------------------------------------------------------
hash#DECIMAL(38, 6)#25%nulls##hash                          7.01ms    142.59
hash#DECIMAL(38, 6)#25%nulls##xxhash64                      7.18ms    139.32
----------------------------------------------------------------------------
hash#DOUBLE#0%nulls##hash                                   3.17ms    315.63
hash#DOUBLE#0%nulls##xxhash64                               2.56ms    391.03
----------------------------------------------------------------------------
hash#DOUBLE#25%nulls##hash                                  3.14ms    318.53
hash#DOUBLE#25%nulls##xxhash64                              2.89ms    345.62
----------------------------------------------------------------------------
hash#REAL#0%nulls##hash                                     1.72ms    582.72
hash#REAL#0%nulls##xxhash64                                 2.58ms    388.13
----------------------------------------------------------------------------
hash#REAL#25%nulls##hash                                    2.74ms    365.21
hash#REAL#25%nulls##xxhash64                                2.70ms    370.39
----------------------------------------------------------------------------
hash#ROW<f_map:MAP<INTEGER,VARCHAR>,f_array:ARR            38.75ms     25.81
hash#ROW<f_map:MAP<INTEGER,VARCHAR>,f_array:ARR            37.36ms     26.77
----------------------------------------------------------------------------
hash#ROW<f_map:MAP<INTEGER,VARCHAR>,f_array:ARR            26.81ms     37.30
hash#ROW<f_map:MAP<INTEGER,VARCHAR>,f_array:ARR            26.36ms     37.94
----------------------------------------------------------------------------
hash#SMALLINT#0%nulls##hash                                 1.66ms    602.42
hash#SMALLINT#0%nulls##xxhash64                             2.36ms    423.71
----------------------------------------------------------------------------
hash#SMALLINT#25%nulls##hash                                2.49ms    402.20
hash#SMALLINT#25%nulls##xxhash64                            2.45ms    408.66
----------------------------------------------------------------------------
hash#TIMESTAMP#0%nulls##hash                                4.73ms    211.51
hash#TIMESTAMP#0%nulls##xxhash64                            4.82ms    207.37
----------------------------------------------------------------------------
hash#TIMESTAMP#25%nulls##hash                               4.19ms    238.73
hash#TIMESTAMP#25%nulls##xxhash64                           4.05ms    246.82
----------------------------------------------------------------------------
hash#TINYINT#0%nulls##hash                                  1.54ms    647.26
hash#TINYINT#0%nulls##xxhash64                              2.24ms    446.65
----------------------------------------------------------------------------
hash#TINYINT#25%nulls##hash                                 2.49ms    401.50
hash#TINYINT#25%nulls##xxhash64                             2.65ms    377.60
----------------------------------------------------------------------------
hash#UNKNOWN#0%nulls##hash                                  1.32ms    757.52
hash#UNKNOWN#0%nulls##xxhash64                              1.28ms    781.78
----------------------------------------------------------------------------
hash#UNKNOWN#25%nulls##hash                                 1.31ms    763.78
hash#UNKNOWN#25%nulls##xxhash64                             1.28ms    778.86
----------------------------------------------------------------------------
hash#VARCHAR#0%nulls##hash                                 11.00ms     90.94
hash#VARCHAR#0%nulls##xxhash64                              9.95ms    100.51
----------------------------------------------------------------------------
hash#VARCHAR#25%nulls##hash                                 9.23ms    108.36
hash#VARCHAR#25%nulls##xxhash64                             8.40ms    119.01
----------------------------------------------------------------------------
```